fix(aarch64): Resolve NotImplementedError during kernel connection

Vamsiindugu opened the pull request:
This PR resolves the `NotImplementedError` that occurred during remote kernel debugging on aarch64.

The error was observed when using `target remote 0:1234` to connect to a QEMU instance. This fix implements the necessary logic to allow the connection to proceed without crashing.

Closes #3271

Vamsiindugu committed:
fix(aarch64): Resolve NotImplementedError during kernel connection

Vamsiindugu commented:
Hello @patryk4815,

I've submitted this PR to address issue #3271 . I've tested the fix on my local setup and confirmed that it resolves the `NotImplementedError` during aarch64 kernel connection.

Please let me know if there are any questions or if you need any further information from my side.

Thank you for your time and consideration.

Vamsiindugu committed:
Trigger CI

patryk4815 commented:
@jxuanli can you review it?

jxuanli commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -472,7 +472,10 @@ def page_shift(self) -> int:
             return 16
-        raise NotImplementedError()
+        elif self.tcr_el1["TG1"] == 0b00:
+            # 0b00 is implementation defined, assume 4KB page size
```
Does Arm have a doc on this?

jxuanli commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -472,7 +472,10 @@ def page_shift(self) -> int:
+        elif self.tcr_el1["TG1"] == 0b00:
+            # 0b00 is implementation defined, assume 4KB page size
+            return 12
+        raise NotImplementedError(f"Unknown TG1 value: {self.tcr_el1['TG1']}")
```
this is currently unreachable because `TG1` is 2 bits

jxuanli commented:
@Vamsiindugu thanks for the PR! I added some comments. We should also change that for userland as well (search for `TG0`).

Vamsiindugu committed:
fix(aarch64): Address review comments for TG0/TG1 and add documentation

Vamsiindugu commented:
Hello @patryk4815 and @jxuanli ,

Thank you for the thorough review. I have carefully read the comments and I am working on the requested changes.

TG1 Documentation: I've located the official ARM documentation which confirms that the 0b00 value for TG1 is reserved. The hardware's behavior is implementation-defined in this case, and the assumption of 4KB is a reasonable choice. I will add a comment to the code referencing this and explaining the logic.

TG0 Handling: This was a great catch. I've also found that TG0's 0b11 value is reserved. I will update the page_shift_user function to handle this case, making the code more robust.

The changes will be pushed to this PR shortly. I appreciate the feedback.

Vamsiindugu committed:
Re-trigger CI to confirm linting fix

jxuanli commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -467,23 +467,39 @@ def ksize(self):
+            # Per ARM ARMv8-A, we assume the hardware will choose a default
+            # implemented size, and we assume 4KB.
+            return 12
+        # NOTE: This line is now unreachable as TG0 is a 2-bit field
```
I think we should just remove those unreachable codes. The the original error was there because not all cases

jxuanli commented:
@Vamsiindugu see my comment above. For linting, just run `./linter.sh -f`. This should fix the issue. Otherwise, this seems a reasonable interpretation of the Arm doc though at some point we may need to refine it further as "implementation defined" does not always mean it is 12. Maybe someone more experienced with Arm can comment on this as well

jxuanli commented:
A more foolproof way of handling this is to find a function that contains information on the size of a page. For example, `copy_page_to_iter` contains info on the size of the page (see `copy_page_to_iter + 108`):
```
pwndbg> disass copy_page_to_iter
Dump of assembler code for function copy_page_to_iter:
   0xffffa070b1fcbe20 <+0>:     paciasp
   0xffffa070b1fcbe24 <+4>:     stp     x29, x30, [sp, #-80]!
   0xffffa070b1fcbe28 <+8>:     str     x25, [sp, #16]
   0xffffa070b1fcbe2c <+12>:    stp     x24, x23, [sp, #32]
   0xffffa070b1fcbe30 <+16>:    stp     x22, x21, [sp, #48]
   0xffffa070b1fcbe34 <+20>:    stp     x20, x19, [sp, #64]
   0xffffa070b1fcbe38 <+24>:    mov     x29, sp
   0xffffa070b1fcbe3c <+28>:    add     x8, x2, x1
   0xffffa070b1fcbe40 <+32>:    mov     x19, x3
   0xffffa070b1fcbe44 <+36>:    mov     x20, x2
   0xffffa070b1fcbe48 <+40>:    cmp     x8, x2
   0xffffa070b1fcbe4c <+44>:    b.cc    0xffffa070b1fcbe58 <copy_page_to_iter+56>  // b.lo, b.ul, b.last
   0xffffa070b1fcbe50 <+48>:    cmp     x8, #0x1, lsl #12
   0xffffa070b1fcbe54 <+52>:    b.ls    0xffffa070b1fcbe9c <copy_page_to_iter+124>  // b.plast
   0xffffa070b1fcbe58 <+56>:    ldr     x9, [x0, #8]
   0xffffa070b1fcbe5c <+60>:    sub     x10, x9, #0x1
   0xffffa070b1fcbe60 <+64>:    tst     w9, #0x1
   0xffffa070b1fcbe64 <+68>:    csel    x9, x0, x10, eq // eq = none
   0xffffa070b1fcbe68 <+72>:    sub     x10, x0, x9
   0xffffa070b1fcbe6c <+76>:    add     x8, x8, x10, lsl #6
   0xffffa070b1fcbe70 <+80>:    cmp     x8, x20
   0xffffa070b1fcbe74 <+84>:    b.cc    0xffffa070b1fcbf30 <copy_page_to_iter+272>  // b.lo, b.ul, b.last
   0xffffa070b1fcbe78 <+88>:    ldr     x10, [x9]
   0xffffa070b1fcbe7c <+92>:    tbnz    w10, #6, 0xffffa070b1fcbe88 <copy_page_to_iter+104>
   0xffffa070b1fcbe80 <+96>:    mov     x9, xzr
   0xffffa070b1fcbe84 <+100>:   b       0xffffa070b1fcbe8c <copy_page_to_iter+108>
   0xffffa070b1fcbe88 <+104>:   ldrb    w9, [x9, #64]
   0xffffa070b1fcbe8c <+108>:   mov     w10, #0x1000                    // #4096
   0xffffa070b1fcbe90 <+112>:   lsl     x9, x10, x9
   ```
   
heuristically determining the page size is perhaps more accurate when functions symbols exist. See `class ArchSymbols`  for more details on this.

Vamsiindugu committed:
fix(aarch64): Address review comments and remove unreachable code

Vamsiindugu committed:
fix(aarch64): Address mypy errors by using a robust page size lookup

Vamsiindugu committed:
fix(aarch64): Use robust page size lookup to fix mypy and runtime errors

Vamsiindugu committed:
fix(aarch64):  fix mypy and runtime errors

Vamsiindugu committed:
fix(aarch64):  fix mypy, runtime errors.

disconnect3d commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -466,24 +466,63 @@ def ksize(self):
+            # 0x1000 is 4096 bytes, which corresponds to a page_shift of 12 (2^12).
+            # The count=100 disassembles the first 100 instructions.
+            for insn in pwndbg.disasm.capstone.disassemble(addr, count=100):
+                if insn.mnemonic == "mov" and "#0x1000" in insn.op_str:
```
The `"#0x1000" in insn.op_str` will also evaluate to `True` for `mov w10, #0x10000` which is probably not what we want.

Can we handle this better somehow?

disconnect3d replied:
cc: @jxuanli @patryk4815 @gsingh93 - I have some mixed feelings about this logic in general (of checking instructions ;/) but maybe its ok as a heuristic?

disconnect3d replied:
Another thing I am worried here is `count=100` - what if this function is extremely simple like: `br some_wrapper_function` and then there is another function just after it with a similar code?

It is probably unlikely, but still, any way to handle this safer?

jxuanli replied:
heuristics should only be used as backups when the values of the fields are reserved and we cannot easily determine the page size. The TG0 and TG1 should otherwise correctly indicate the page size. @Vamsiindugu sorry that I didn't make that clear, could you change that? you can also use `has_debug_symbols` or `requires_debug_symbols` (which has the option to return a default value) to check if the symbol exist

jxuanli replied:
so, if we encounter reserved values, we check heuristics. And if heuristics fails, we either make it panic or return a default value

disconnect3d commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -466,24 +466,63 @@ def ksize(self):
+            # sizes. We assume the common 4KB page size.
+            # Reference: ARM ARMv8-A, TCR_EL1 register description.
+            return 12
+            # This will satisfy mypy because the function will always return a value.
```
```suggestion
        # This will satisfy mypy because the function will always return a value.
```

disconnect3d committed:
Update pwndbg/aglib/kernel/paging.py

disconnect3d commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -466,24 +466,63 @@ def ksize(self):
-            return 14
-        raise NotImplementedError()
+            return 14  # 16KB granule
+        elif self.tcr_el1["TG0"] == 0b11:
```
Shall we maybe use `match self.tcr_el1["TG0"]` instead of if elif?

disconnect3d commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -466,24 +466,63 @@ def ksize(self):
-            return 16
-        raise NotImplementedError()
+            return 16  # 64KB granule
+        elif self.tcr_el1["TG1"] == 0b00:
```
Shall we maybe use `match self.tcr_el1["TG0"]` instead of if elif?

disconnect3d commented:
> A more foolproof way of handling this is to find a function that contains information on the size of a page. For example, `copy_page_to_iter` contains info on the size of the page (see `copy_page_to_iter + 108`):
> 
> ```
> pwndbg> disass copy_page_to_iter

Just to be sure: which kernel version (or blob) is this disasm from?

Vamsiindugu committed:
Merge branch 'dev' into fix/aarch64-kernel-connect

Vamsiindugu committed:
fix(aarch64): Improve page size detection and address review feedback

Resolves NotImplementedError during aarch64 remote kernel debugging by enhancing page size detection.

- Implements a more robust heuristic to determine page size by disassembling the `copy_page_to_iter` kernel function. This approach is now safer against false positives and checks for multiple page sizes (4KB, 16KB, 64KB).
- Refactors `page_shift` and `page_shift_user` methods to use `match` statements for improved readability, as suggested by reviewers.
- Fixes various linting errors.

Vamsiindugu committed:
Merge branch 'fix/aarch64-kernel-connect' of https://github.com/Vamsiindugu/pwndbg into fix/aarch64-kernel-connect

Vamsiindugu committed:
fix(aarch64): Update symbol lookup method for page size detection

Vamsiindugu committed:
Merge branch 'dev' into fix/aarch64-kernel-connect

Vamsiindugu committed:
fix(aarch64): Update symbol lookup method for disassembly in Aarch64PagingInfo

Vamsiindugu committed:
Merge branch 'fix/aarch64-kernel-connect' of https://github.com/Vamsiindugu/pwndbg into fix/aarch64-kernel-connect

Vamsiindugu committed:
fix(aarch64): Correct disassembly import path in Aarch64PagingInfo

Vamsiindugu committed:
fix(aarch64): Update disassembly function call in Aarch64PagingInfo

jxuanli commented:
@disconnect3d 6.10.9 but the version shouldn't really matter here as I believe the definition has been fairly consistent https://elixir.bootlin.com/linux/v6.16.4/source/lib/iov_iter.c#L349.

Vamsiindugu committed:
feat(aarch64): Enhance Aarch64PagingInfo with page size and shift properties

Vamsiindugu committed:
Revert "feat(aarch64): Enhance Aarch64PagingInfo with page size and shift properties"

This reverts commit 81bb7b75fde6d95d1c8db771f7b80f3c038042c0.

disconnect3d commented on pwndbg/aglib/kernel/paging.py:
```diff
@@ -466,24 +466,78 @@ def ksize(self):
+            # Common page sizes are 4KB (0x1000), 16KB (0x4000), 64KB (0x10000).
+            page_sizes = {0x1000: 12, 0x4000: 14, 0x10000: 16}
+            # The count=100 disassembles the first 100 instructions. This is a heuristic.
+            for insn in pwndbg.disasm(addr, count=100):
```
Seems like a bug?

disconnect3d replied:
Also @jxuanli suggested we do this as last resort iiuc
